src/test/python/deep_qa_test/layers/attention/attention_test.py

```diff
@@ -1,13 +1,92 @@
 # pylint: disable=no-self-use,invalid-name
 
-import numpy
+import numpy as np
 from numpy.testing import assert_almost_equal
 from keras.layers import Embedding, Input
 from keras.models import Model
-
+import keras.backend as K
 from deep_qa.layers.attention.attention import Attention
 
 class TestAttentionLayer:
+    # pylint: disable=no-self-use
+    def test_no_mask(self):
+        vector_length = 3
+        matrix_num_rows = 2
+
+        vector_input = Input(shape=(vector_length,),
+                             dtype='float32',
+                             name="vector_input")
+        matrix_input = Input(shape=(matrix_num_rows, vector_length),
+                             dtype='float32',
+                             name="matrix_input")
+        similarity_softmax = Attention()([vector_input, matrix_input])
+        model = Model([vector_input, matrix_input],
+                      similarity_softmax)
+
+        # Testing general non-batched case.
+        vector = np.array([[0.3, 0.1, 0.5]])
+        matrix = np.array([[[0.6, 0.8, 0.1], [0.15, 0.5, 0.2]]])
+
+        result = model.predict([vector, matrix])
+        assert_almost_equal(result, np.array([[0.52871835, 0.47128162]]))
+
+        # Testing non-batched case where inputs are all 0s.
+        vector = np.array([[0, 0, 0]])
+        matrix = np.array([[[0, 0, 0], [0, 0, 0]]])
+
+        result = model.predict([vector, matrix])
+        assert_almost_equal(result, np.array([[0.5, 0.5]]))
+
+    def test_masked(self):
+        # Testing general masked non-batched case.
+        vector = K.variable(np.array([[0.3, 0.1, 0.5]]))
+        matrix = K.variable(np.array([[[0.6, 0.8, 0.1], [0.15, 0.5, 0.2], [0.1, 0.4, 0.3]]]))
+        mask = K.variable(np.array([[1.0, 0.0, 1.0]]))
+        result = K.eval(Attention().call([vector, matrix], mask=["_", mask]))
```

Comment:
im not sure this is a good idea actually, maybe keras does some stuff internally that we miss out on when we just test __call__ directly. At least it was the case for me (with the whole not a keras tensor thing) that these tests would pass, but the others that actually used a model and the full framework would fail which seems like they do something abstracted away from the user. Besides, having them in the whole model scaffolding is more accurate of their actual use, right? The reason i did this is because the situation with masking layer only masking if the element at the final dimension is the same value, and since we have arrays and they take floats it isn't really compatible...